### ISSUE TITLE
[BUG] (Terribly) wrong sizing for large subplots 

### ISSUE THREAD

**Issue by roland-KA** [JuliaPlots/Plots.jl#4046]
When creating a large subplot (8 x 3 diagrams in my case), Plots gets completely confused in determining the correct size of the whole plot:
- Without specifying `size`, the result is a complete mess of overlapping subplots. When reducing the number of subplots (eg. to 2 x 3 diagrams), the issue disappears.
- But even when specifying `size`, depending on the specific parameters used, parts of the whole plot (e.g. ticks on the left or bottom) are not visible.

So there are two questions:
- How can I get a correct automated sizing of the complete plot (e.g. with 24 subplot)?
- If an automated sizing is not possible, how can I determine `size` parameters (not using 'try & see') so that all elements of the plot are visible?

These problems may be related to #3378 and #1909.

In the following 'Details' sections the issues are shown for GR (which I'm using mainly) and PlotlyJS (which has different issues).

### Complete plot without specifying `size`
Here the result with the complete subplot. In this case the following error message appears: `GKS: Rectangle definition is invalid in routine SET_VIEWPORT`

<img width="588" alt="image" src="https://user-images.githubusercontent.com/80126696/149772164-3a546fd8-eea3-4d58-be8a-575a87d13ef4.png">

### Reducing the plot to e.g. 2 x 3 subplots 
When reducing the size of the plot to e.g. 2 x 3 diagrams, the sizing works well.

<img width="588" alt="image" src="https://user-images.githubusercontent.com/80126696/149773228-bf6f4fa1-369a-4621-9b1e-54ed526458ce.png">

### Specifiying `size` with the 'wrong' parameters
Here I've used `size = (1000, 2400)`. In this case the complete plot (i.e. all 8 x 3 subplots) are shown, but the y-ticks of the leftmost diagrams are not visible and the y-ticks of the other two diagrams overlap with their left neighbors  (the pictures shows only the first two rows of the plot):

<img width="629" alt="image" src="https://user-images.githubusercontent.com/80126696/149773911-7c0475f1-acb0-4f35-a307-aec61ee9e322.png">

### Specifying 'correct' `size` parameters
In the following example, I've used `size = (1500,2000)` and everything worked fine (e.g. the y-ticks of the leftmost diagrams are visible):

<img width="629" alt="image" src="https://user-images.githubusercontent.com/80126696/149774943-d2fe28d8-97fe-4b2e-a2c5-e560bc77a118.png">

## Details (using PlotlyJS)
I've only shortly tested the examples above using PlotlyJS. This results in different issues:

- Without specifying `size` the whole plot is created, but resulting in tiny subplots and PlotlyJS is apparently trying to create a 'unified' legend for all the histograms in the right column, which isn't the desired outcome.

<img width="550" alt="image" src="https://user-images.githubusercontent.com/80126696/149778053-d695523c-4ea6-4121-9001-b98b04a35b63.png">

- Reducing the number of subplots results in overlapping subplots and the issue with the 'unified' legend remains:
<img width="550" alt="image" src="https://user-images.githubusercontent.com/80126696/149778212-f0bcefbd-c905-47b4-adcb-abd221166af7.png">

- Specifying `size` seems to create plots which show all details (as PlotlyJS creates larger plots using a scrollable region). But the legend is not correct:
<img width="550" alt="image" src="https://user-images.githubusercontent.com/80126696/149778721-f45dc442-3965-4c80-b85c-e5574e5423b5.png">

### Backends

This bug occurs on ( insert `x` below )

Backend      | yes | no  | untested
-------------|-----|-----|---------
gr (default) |   x  |     |
pyplot       |     |     | x
plotlyjs     |   x  |     | 
pgfplotsx    |     |     | x
unicodeplots |     |     | x
inspectdr    |     |     | x
gaston       |     |     | x

### Versions

Plots.jl version: 1.25.5
Backend  version (`]st -m <backend(s)>`): GR 0.63.1, PlotlyJS 0.18.8
Output of `versioninfo()`:

Julia Version 1.6.1
Commit 6aaedecc44 (2021-04-23 05:59 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, westmere)
Environment:
  JULIA_REVISE_WORKER_ONLY = 1

Alle examples are created within a Pluto notebook (using Pluto version 0.17.5).

**Comment by t-bltg** [comment]
Can you post something reduced (`mwe`) that we can work on ? Let's focus on `GR` for now.

**Comment by roland-KA** [comment]
I've created now a Pluto notebook with a minimal version (using GR):
[Plots-LayoutProblem.jl.zip](https://github.com/JuliaPlots/Plots.jl/files/7905506/Plots-LayoutProblem.jl.zip)
 
In this example it became clear that the error is related to using `xlabel`. If I don't use x-lables on the diagrams, the resulting layout is ok (apart from the fact, that _GR_ and _plotlyjs_ are using different layout strategies; _GR_ puts everything in a rectangle of a pre-defined, fixed size whereas _plotlyjs_ paints the subdiagrams and then calculates the bounding box; but that's another story). 

As soon as x-labels are added to the subdiagrams, the layout of the whole plot doesn't work any more.

**Comment by isentropic** [comment]
Please provide a MWE not a pluto notebook, it will be easier to help this way

**Comment by roland-KA** [comment]
This is the code from the notebook:

```Julia
using Plots
using StatsPlots

m = rand(20, 24)
names = collect("abcdefgh")
violins = [violin(m[:, i], xlabel = names[i % 3 + 1], leg = false) for i in 1:24];
plot(violins..., layout = (8,3))
```

I think it's really minimal now 🤓

**Comment by isentropic** [comment]
I can't really understand why do you consider this a bug. Since the font sizes, marker size, and figure sizes are constant, you're overconstraining the system too much. You're putting to many subplots on a small size figure. You either:
1) increase figure size
2) decrease marker/font size

How should Plots go about overconstrained plots instead?

I get a reasonable result via pyplot:  `plot(violins..., layout = (8,3), size=(2000,2000))`
![image](https://user-images.githubusercontent.com/24591123/150465131-2762c362-d48f-4650-9607-c213fab3b99e.png)

**Comment by isentropic** [comment]
I get same output with GR too

**Comment by roland-KA** [comment]
My original post involved two issues:
1.  How can I get (reasonable) automated sizing?
2.  If I set the `size` parameters manually, the layout algorithm often cuts off some border elements. How can I (at least) determine correct `size` parameters?

I think, your question refers to my first issue. But let's talk about both.

## Automated sizing

There are two questions involved: 
1. Which strategy should be used to layout subplots? To my knowledge there are two different strategies:
    - _Fixed size area_: There is a plotting area of fixed size and the layout algorithm tries to fit all subplots inside this area. 
    - _Variable size area_: The layout algorithm uses a reasonable size for each subplot and calculates on this basis the area needed to display the whole plot.
    
    From my few experiments, I've seen that both strategies are used depending on the graphics package and it's backend:
    - Fixed size: Plots/GR, Plots/plotly on REPL, Plots/plotlyjs on REPL, Gadfly
    - Variable size: Plots/plotly in Pluto, Plots/plotlyjs in Pluto

    Both strategies have pros and cons. But within a package like `Plots` **there should be consistency**. One of the two strategies should be used by all backends in all contexts.

2. If a '_fixed size_' approach is used: How should the algorithm react to tightening constraints (i.e. what should happen, when more and more subplots are added)?

    In this case there should be a so-called **_graceful degradation (as long as possible)_**. That means the layout algorithm should try to fit as many subplots as possible in the available space. Of course the more subplots are added, the smaller they get (and thus the more unreadable). But the user of the package can still see in this situation what happens and where the problem is and adapt parameters like `size` to solve the problem.

    If the layout algorithm 'gives up' und degrades rather ungraceful, as in my example, you get no hint on what's wrong.

    So it's perhaps no 'hard' bug, but also a **question of consistency**. Other backends (like _plotly_ or _plotlyJS_) and other packages like `Gadfly` degrade more graceful. Using them, I could create even diagrams with 12 rows, whereas GR stopped after 7 rows. 

## Setting `size` parameters manually

I think here we have a real bug, since even if I specify an area large enough to fit in all subplots (like `size = (1500, 2000)`, the layout algorithm cuts off often some elements on the borders of the plot ... and it does this *without any warning*!

So it should at least print some warning or (better) use a layout so that everything fits inside the boundaries given. ... and as long as this doesn't work, it would be interesting to know how reasonable `size` parameters can be calculated manually.

**Comment by isentropic** [comment]
Yeah, I understand what you mean, so you basically want `size = :auto` which has been discussed internally too. This might be a long-term feature.

**Comment by isentropic** [comment]
I assume you'd agree with the title and category change

**Comment by roland-KA** [comment]
Thank you for your feedback! It helped me to better understand the issues I brought  up in my original post. 

With the change of the title we address now partly, what I wrote in section '_automated sizing'_. Especially the part where I describe the two possible approaches to automated sizing. These could be selected in a future version of `Plots.jl` e.g. using a parameter.

But there are two aspects, which I don't see addressed by the renamed issue:
1. **GR should behave similar to other backends when plotting large subplots**. I.e. it should output some reasonable plots the same way _plotly_ or _plotlyJS_ do. From a software engineering view point, `Plots` is an interface for different backends. A basic quality requirement for an interface is, that a function call produces the same output, independent of the underlying implementation. This requirement (which is BTW a design goal of Plots, namely 'Consistency') is here violated. So it should be fixed.
2.  **When specifying `size`, Plots should not cut off parts of the plot without warning**. As I described, this happens even when `size` specifies an area which should be large enough to show the whole plot.

So I will open two separate issues for these aspects ... or do you have a better idea?

**Comment by t-bltg** [comment]
> So I will open two separate issues for these aspects ... or do you have a better idea?

Please do as 1. is a matter of consistency among backends (which is hard hard to fix since a universal algorithm may not be implementable without underlying changes in the backends), and 2. is a per-backend scaling fix or margins / layout computation issue.

Overall, I do agree with your comments about consistency and graceful degradation. I you have any ideas on implementation, we'd be happy to review some PRs improving the current situation.

**Comment by t-bltg** [comment]
> I assumed that the layout logic would be completely implemented in Plots.

Yes, (see https://docs.juliaplots.org/latest/layouts/ or https://github.com/JuliaPlots/Plots.jl/blob/311831ca4eb612bc2db0cc88249aa83f07292e3f/src/layouts.jl#L530), however the backends do behave differently on scaling (different fonts, glyphs, ...).

**Comment by iabraham** [comment]
I am having a [similar issue.](https://stackoverflow.com/questions/75893295/julia-plots-xtick-and-ytick-font-halign-and-valign). Did anyone figure out workarounds or fixes for this?

 More **specifically**, in roland-KA's _Specifiying size with the 'wrong' parameters_ figure, one can see that the ytick labels are far away from the y-axis. In my experiments as more subplots are added to a figure, somehow there seems to be a minimum spacing between the labels and the axis that creeps in. 

I have been unable to change that spacing with any of the margin options and it doesn't seem to change with the figure size but rather looks tied to the number of subplots.
